helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java

```diff
@@ -519,6 +520,8 @@ protected void handleJobTimeout(JobContext jobCtx, WorkflowContext workflowCtx,
     _clusterStatusMonitor.updateJobCounters(jobCfg, TaskState.TIMED_OUT);
     _rebalanceScheduler.removeScheduledRebalance(jobResource);
     TaskUtil.cleanupJobIdealStateExtView(_manager.getHelixDataAccessor(), jobResource);
+    // New pipeline trigger for workflow status update
+    RebalanceUtil.scheduleOnDemandPipeline(_manager.getClusterName(),0L,false);
```

Comment:
2 questions.
1. I assume the IS change will trigger the pipeline. But obviously, if the TF logic says goodbye to IS, then this won't trigger the right pipeline anymore.
2. In this case, shall we just listen to the context change? Depends on onDemond pipeline is not a scalable solution in general.

Please correct me if I misunderstood the first point.

Reply:
Your understanding is correct: that's why we need the on-demand trigger now without IS removal event. 

About the second point: these are the only places IS removal events are necessary to trigger rebalance. It doesn't need to scale up. 

@alirezazamani Could you chime in here on @jiajunwang 's proposal? I think context-based rebalance might be an overkill and introduce problems we are not ready to face. (Either way I don't think something like that should be tackled in this PR)

Reply:
I think this will be overkill as well and not necessary. Basically, I am looking the context as controller's output and reacting on output is not ideal from my prospective. Also, unlike regular resources, context changes are much more frequent in TF resources. So there will be many redundant pipeline runs and can cause other issue. In my opinion, we can stick to ondemand rebalancer for these corner cases.

Reply:
Then, can we add scheduleOnDemandPipeline into cleanupJobIdealStateExtView() ? Otherwise, cleanupJobIdealStateExtView is not a complete call that triggers the expected rebalance, right?
It also helps to simplify code, IMO.

Reply:
`cleanupJobIdealStateExtView()` is kept for backward-compatibility purpose and is marked as deprecated now; it will be removed on a later phase. I suppose we put the on-demand rebalance in `cleanupJobIdealStateExtView()` now, and move it out when we delete `cleanupJobIdealStateExtView()`. Do you think that's necessary @jiajunwang ? Seems not to be too much difference.

Reply:
This is necessary. So no matter who does the cleanup later, he or she won't forget to remove the scheduleOnDemandPipeline(). High-level, these 2 logics are bundled. And we shall not assume everyone remembers this relationship by their memory or document. Organize the code to ensure it.

Reply:
As we discussed, please add a TODO there for the future change. For now, it is OK to be left here.